Add memory format support to `resize_as_` operator

VitalyFedyunin opened the pull request:
Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29461 [WIP] Experiments
* #29460 [WIP] Switch default of clone, to, *_like ops
* #29459 [WIP] Prepare copy op to switch memory format preservation defaults
* #28991 Adding function to convert Module to channels last
* #28544 Add torch.memory_format support to the TorchScript
* #28292 Add memory format support to the `resize_` op.
* #28291 TensroIterator preserve format for binary, ternary operators.
* #28077 Add channels last support to cuda.comm.scatter and gather
* #28076 Kill `operator==` of TensorOptions as confusing one
* **#27979 Add memory format support to `resize_as_` operator**

Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

VitalyFedyunin committed:
Add memory format support to `resize_as_` operator

Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

zou3519 commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -21,15 +22,26 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
+    const Tensor& the_template,
+    c10::optional<MemoryFormat> optional_memory_format) {
   if (self.is_sparse() && the_template.is_sparse()) {
     return native::resize_as_sparse_(self, the_template);
```
It seems weird that we're just ignoring the memory format for sparse. Would it be better to assert that optional_memory_format is nullopt?

VitalyFedyunin replied:
done

VitalyFedyunin committed:
Rebase on top of master on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Fixing casting ops and tests on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Rebasing to master on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

[ghstack-poisoned]

VitalyFedyunin committed:
Rebasing to master on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Update on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin commented:
@ailzhang ideally I want to land all this stack Monday morning

VitalyFedyunin committed:
Fix rand_like test on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ linter on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix onnx on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Rebase on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like not simple mapable any more on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
randn_like mappable again and supported in jit with memory format on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -21,15 +22,29 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
+  auto memory_format =
+      optional_memory_format.value_or(MemoryFormat::Contiguous);
   if (self.is_sparse() && the_template.is_sparse()) {
+    TORCH_CHECK(
```
@zou3519 suggestion doesn't seem done -- you can't do it without checking if a value is set and you are calling `value_or`.

VitalyFedyunin replied:
how about now

gchanan replied:
yes, looks correct now.

gchanan commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -21,15 +22,29 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
     return native::resize_as_sparse_(self, the_template);
   }
   Tensor& result = self.resize_(the_template.sizes());
+  if (memory_format == MemoryFormat::Preserve) {
```
documentation?

It's not obvious to me how memory_format even applies to resize_, given the crazy semantics of resize.  Can you document all the behaviors (sharing X non_contiguous X larger/smaller) and what this means?

I suspect we are only trying to support some subcases of this.

VitalyFedyunin replied:
I will put documentation, but generally speaking I had to convert this function as some models and developers is using it similar to empty_like, but resizing existing tensor. So to make this models BC compatible we need to preserve memory format similarly.

gchanan replied:
I get that but I'm not convinced we want the specific implementation here.

VitalyFedyunin replied:
Updated in the PR description

VitalyFedyunin committed:
fix bc test for empty_like on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
fix bc test for empty_like now with gradient on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
rebase to master on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
add TensroIterator tests on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Add some tests on "Add memory format support to `resize_as_` operator"


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---
Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Addressed PR feedback on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix Autograd on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -21,15 +22,29 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
+    TORCH_CHECK(
+        !optional_memory_format.has_value(),
+        "Unsupported memory format for sparse tensor resize_as_ :",
+        optional_memory_format.value_or(MemoryFormat::Contiguous));
```
can't you use `optional_memory_format.value()` or `*optional_memory_format`, because you know it has a value?

gchanan commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -21,15 +22,29 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
+  if (memory_format == MemoryFormat::Preserve) {
+    memory_format = the_template.suggest_memory_format();
+  }
+  self.unsafeGetTensorImpl()->empty_tensor_restride(memory_format);
```
doesn't this change the behavior, even in cases where there's no channel_last tensors?  i.e. `non_contig.resize_as_(contig.shape)` will make a tensor contiguous when it didn't previously.

Is that what you want?

VitalyFedyunin replied:
Generally line 38 above `Tensor& result = self.resize_(the_template.sizes());` was already setting up strides to contiguous. 

But thank to this comment I found the only case when it is not intended and might be BC breaking:

```python
x =  torch.empty(2,3).t()
x.resize_as_(x)
# x.is_contiguous() == True but supposed to keep original strides.
```

Gonna fix it now and test cover (we had no such test).

VitalyFedyunin committed:
Fix resize_as_ with new test on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix resize_as_ on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

gchanan reviewed (changes requested):
the behavior looks correct but this badly needs documentation.

VitalyFedyunin committed:
Add resize_as_ docs on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin commented:
Docs added

gchanan commented on aten/src/ATen/native/Resize.cpp:
```diff
@@ -21,15 +22,51 @@ Tensor& resize_cpu_(Tensor& self, IntArrayRef size) {
+// Strides of the output tensor of `resize_as_` operator is defined by input
+// tensor strides and the value of memory_format argument.
+//
+// If memory_format is not defined or equals to MemoryFormat::Contiguous
```
I don't think this is correct.

Don't we not restride if memory_format is not set and the sizes match?

VitalyFedyunin replied:
Fixed

VitalyFedyunin committed:
fix resize_as_ docs on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

VitalyFedyunin committed:
fix resize_as_ docs on "Add memory format support to `resize_as_` operator"

Rationale to add optional memory format to `resize_as_` and `resize_` operators is the fact that they are frequently used as inplace `empty_like` and `empty` operators inside of our code base. So having them to accept memory format similarly to `empty_like` and `empty` seems to be logical.

We could also add two new operators, but it will be more confusing (taking into account that renaming existing operators in not an option).

--- 


Adds memory_format keyword argument (positional for cpp).

'Preserve' behavior now follows next rules:
1) If tensor is non-overlapping and dense - output tensor will have the same strides as input tensor.
2) If not (1) and tensor is stored in the channels last format, output tensor going to have channels last format.
3) Output tensor is going to be contiguous in all other cases.

---

Dense tensor is the tensor that store values in a contiguous block of memory.
Non-overlapping tensor is the tensor in which elements occupy individual non-repetitive memory.

Differential Revision: [D17980311](https://our.internmc.facebook.com/intern/diff/D17980311)

[ghstack-poisoned]

gchanan commented on test/test_torch.py:
```diff
@@ -11963,6 +11963,18 @@ def test_memory_format_preserved_after_permute(self, device):
+        x.resize_as_(x)
+        self.assertEqual(x.stride(), old_strides)
+
+    def test_memory_format_resize_as(self, device):
```
we should have tests for all the cases, i.e. add a:
1) explicit contiguous test
2) explicit channels_last test.
